enclave/core/sgx/cpuid.c

```diff
@@ -20,27 +20,26 @@ static uint32_t _cpuid_table[OE_CPUID_LEAF_COUNT][OE_CPUID_REG_COUNT];
 **
 **==============================================================================
 */
-oe_result_t oe_initialize_cpuid(uint64_t arg_in)
+oe_result_t oe_initialize_cpuid(oe_init_enclave_args_t* args)
 {
     oe_result_t result = OE_UNEXPECTED;
-    oe_init_enclave_args_t* args = (oe_init_enclave_args_t*)arg_in;
-    if (args != NULL)
-    {
-        // Abort the enclave if AESNI support is not present in the cached
-        // CPUID Feature information (cpuid leaf of 1)
-        if (!(args->cpuid_table[1][OE_CPUID_RCX] & OE_CPUID_AESNI_FEATURE))
-            oe_abort();
 
-        OE_CHECK(oe_memcpy_s(
-            _cpuid_table,
-            OE_CPUID_LEAF_COUNT * OE_CPUID_REG_COUNT *
-                sizeof(_cpuid_table[0][0]),
-            args->cpuid_table,
-            OE_CPUID_LEAF_COUNT * OE_CPUID_REG_COUNT *
-                sizeof(args->cpuid_table[0][0])));
+    if (args == NULL || !oe_is_within_enclave(args, sizeof(*args)))
```

Comment:
Side note: The null check on `args` is only necessary because of the `sizeof(*args)` construction here (`oe_is_within_enclave` already rejects null pointers) and it looks like `_handle_init_enclave` gets around this by using `sizeof(oe_init_enclave_args_t)` to elide the double check.

We should probably consider standardizing at some point.

Reply:
@CodeMonkeyLeet 
Though redundant, I added the pre-conditions to this function to keep this function safe from future modifications to the caller.

Note: In C/C++, the arguments to sizeof expressions is never evaluated. You could do sizeof(printf("Hello, World\n")); the language guarantees that the printf is used only for inferring the type. So there is no need to guard sizeof(*args) with a NULL check.

I personally prefer sizeof(expression) where possible instead of sizeof(type) since that is robust to expressions changing type in the future.